### DISCUSSION TITLE
WebSocket library

### DISCUSSION THREAD

**Discussion by davidbrochart** [frankie567/httpx-ws#6] [General]
Hi @frankie567, interesting project!
Out of curiosity, why did you choose `wsproto` over [websockets](https://github.com/aaugustin/websockets)?

**Comment by frankie567** [comment]
Thanks David! Well, actually I followed what Tom Christie recommended in this thread: https://github.com/encode/httpx/issues/304#issuecomment-1320023905 

Since I trust him to know way better than me what is better, I didn't question the choice 😬

**Comment by davidbrochart** [comment]
Thanks for the answer, that makes sense.
I'll probably use `httpx-ws` very soon :smiley:

**Comment by Kludex** [comment]
@aaugustin Do you have a list of pros and cons about `wsproto` and `websockets` regarding the sans-io implementation?

Asking here because it may give some insights... :man_shrugging:

**Comment by aaugustin** [comment]
In terms of API, websockets does more for you:

- It handles ping / pong for you — while wsproto asks you to do it
- It handles the closing handshake for you — same comment
- It tells you when you should close the TCP connection — the WebSocket spec requires the server to close the connection first, which causes a surprising amount of pain when you implement the closing handshake on the client side — and that drove some aspects of the design of websockets' Sans-I/O API.

In terms of performance, you should see a big difference in favor of websockets when dealing with large frames (like, 1MB) because [this code in wsproto](https://github.com/python-hyper/wsproto/blob/a7d9ff0c34567b042d025db555595bb6234f89e8/src/wsproto/frame_protocol.py#L21-L42), even though it's reasonably optimized Python, is nowhere near as fast as [that code in websockets](https://github.com/aaugustin/websockets/blob/cac6e8575e7b3c6339817c74fcbba846a4f935dc/src/websockets/speedups.c#L127-L179). This is actually a big performance bottleneck of all pure Python implementations.

I believe that websockets a bit more battle-tested as it predates wsproto by a few years — some super-edge-case bugs were ironed out over the years.

You're welcome to have Opinions about websockets' regex-based HTTP implementation. Well, it's fast, I cannot remember having to fix embarrassing bugs, and OSS-Fuzz didn't find anything after running billions of test cases. So I guess it's a viable way implement the required subset of HTTP for a WebSocket handshake? Anyway, it doesn't matter here, as you're planning to use httpx for parsing HTTP messages.

[The documentation of websockets' Sans-I/O layer](https://websockets.readthedocs.io/en/stable/howto/sansio.html) tells you the basics but isn't at the same level as the documentation of the rest of the library — because so few people will integrate the Sans-I/O layer that I can support them individually. It's probably more useful to check this PR: https://github.com/aaugustin/websockets/pull/1245/files (or the actual code in the library once I merge it).

There may be other aspects in which wsproto is superior to websockets. I know where I put my efforts in websockets; I don't know where the wsproto crew invested!

Beyond that, WebSocket is a relatively small protocol so you will likely find pretty much the same code. At some point, websockets had higher performance; then everyone adopted the same ideas and I'm not sure there's a

**Comment by Kludex** [comment]
Thanks @aaugustin :pray:
